actionview/lib/action_view/renderer/partial_renderer/collection_caching.rb

```diff
@@ -0,0 +1,87 @@
+require 'active_support/core_ext/object/try'
+
+module ActionView
+  module CollectionCaching # :nodoc:
+    extend ActiveSupport::Concern
+
+    included do
+      # Fallback cache store if Action View is used without Rails.
+      # Otherwise overriden in Railtie to use Rails.cache.
+      mattr_accessor(:collection_cache) { ActiveSupport::Cache::MemoryStore.new }
+    end
+
+    private
+
+    def cache_collection_render
+      return yield unless cache_collection?
+
+      keyed_collection = collection_by_cache_keys
+      partial_cache = collection_cache.read_multi(*keyed_collection.keys)
+
+      unless partial_cache.empty?
+        ActionView::Base.logger.try :debug, "#{partial_cache.size} partials " \
+          "multi read from cache."
+      end
+
+      @collection = keyed_collection.reject { |key, _| partial_cache.key?(key) }.values
+      rendered_partials = @collection.any? ? yield.dup : []
+
+      unless rendered_partials.empty?
+        ActionView::Base.logger.try :debug, "#{rendered_partials.size} partials" \
+          " will be written to cache."
+      end
+
+      fetch_or_cache_partial(partial_cache, order_by: keyed_collection.each_key) do
+        rendered_partials.shift
+      end
+    end
+
+    def cache_collection?
+      @options.fetch(:cache, automatic_cache_eligible?)
+    end
+
+    def automatic_cache_eligible?
+      single_template_render? && !callable_cache_key? && template_eligible?
+    end
+
+    def single_template_render?
+      @template # Template is only set when a collection renders one template.
+    end
+
+    def callable_cache_key?
+      @options[:cache].respond_to?(:call)
+    end
+
+    def template_eligible?
+      @template.eligible_for_collection_caching?(as: @options[:as]).tap do
+        ActionView::Base.logger.try :debug,
+          "Collection automatically caches #{@template.virtual_path}, so next render uses a multi read for any templates rendered now."
```

Comment:
I did remember seeing lines containing the fragment, but I don't think they were from read_multi. I'll recheck tomorrow.

I'm also logging the count of partials read from and written to cache. But I don't know how useful that is.

Sorry about this taking so long :)

--
Kasper

> Den 18/02/2015 kl. 22.06 skrev David Heinemeier Hansson <notifications@github.com>:
> 
> In actionview/lib/action_view/renderer/partial_renderer/collection_caching.rb:
> 
> > +    def automatic_cache_eligible?
> > +      single_template_render? && !callable_cache_key? && template_eligible?
> > +    end
> > +
> > +    def single_template_render?
> > +      @template # Template is only set when a collection renders one template.
> > +    end
> > +
> > +    def callable_cache_key?
> > +      @options[:cache].respond_to?(:call)
> > +    end
> > +
> > +    def template_eligible?
> > +      @template.eligible_for_collection_caching?(as: @options[:as]).tap do
> > +        ActionView::Base.logger.try :debug,
> > +          "Collection automatically caches #{@template.virtual_path}, so next render uses a multi read for any templates rendered now."
> Hmm, I wonder if it won't actually be enough to rely on the multi_get render call since we'll see lines like "Read fragment views/recordings/431-6/dc02965b56829c216cdad6b48ecfaabc". Not sure what the multi_get lines logs.
> 
> —
> Reply to this email directly or view it on GitHub.
>